### ISSUE TITLE
Pass a tensor of `-inf` to `nn.Softmax` and return a tensor of `nan`

### ISSUE THREAD

**Issue by zhangguanheng66** [pytorch/pytorch#25110]
# Issue description

Just to check. When passing a tensor of `-inf` to `nn.Softmax`, it returns a tensor of `nan`. I know it's a very rare case but do we expect this result? Should we add a warning there?
Relevant issue https://github.com/pytorch/pytorch/issues/24816

## Code example

Please try to provide a minimal example to repro the bug.

```
import torch
x = torch.Tensor([[[float("-inf"), float("-inf"), float("-inf")]]])
softmax = torch.nn.Softmax(dim=-1)
softmax(x)
>>> tensor([[[ nan, nan, nan]]])
```

**Comment by SsnL** [comment]
0/0 is NaN. I feel this is expected.

**Comment by zhangguanheng66** [comment]
Thanks @SsnL . It would be great if there is a warning in this case.

**Comment by izdeby** [comment]
Not a bug. Closing the issue. @zhangguanheng66, feel free to reopen it if you think that warning should be present.

**Comment by soumith** [comment]
> It would be great if there is a warning in this case.

Generally not possible without a full pass through the data, which is perf overhead

**Comment by harryhan618** [comment]
@soumith 
Hi, May I ask '-inf' in softmax is numerical stable?

Here's my situation:
Say, I have `x` of shape (nB, nC, H, W) and a `0/1` `mask` same shape.
I want those masked position won't get involved in `softmax` at dimension `nC`:
``` python
with torch.no_grad():
    x[mask] = float('-inf')
x = torch.softmax(x, dim=1)
```

I wanna know if this should be a proper way.
Thanks!

**Comment by arui-yyz** [comment]
How does torch handle nn.softmax with -inf input? It looks like torch does not output NaN anymore in this case.

**Comment by nikitaved** [comment]
Softmax stills produces `nan`s in such cases. If `-inf` is assumed to be in the limit, then the result should be a uniform distribution, if not, then 0/0 kills it. PyTorch computes stable `softmax(x)` by computing `softmax(x - x.min())` instead. But `x.min() = -inf`, will automatically imply uniformity, so this case is very easy to detect... What do you think?
